<a href="https://colab.research.google.com/github/mrsferret/Code-Division/blob/main/Worksheets/10_2_Movies_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Movies Mini-project
---

In the previous worksheet you converted an SQL relational database to a single pandas dataframe and downloaded it. You will be analysing it today.

If you were unable to download the file, there is a copy located here: 

"https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"



### Inspect the dataset 
---

In [81]:
import pandas as pd

def inspect_data(df):

  # firstly, sort the data on movie, director and year
  df = df.sort_values(by = ["name", "last_name", "first_name", "year"])
  display(df.describe())
  display(df.head())
  display(df.info())

  # check the data for NA/NaN/missing values across the dataframe
  display("Null Values: ", df.isnull().values.any())

  #============================================================================
  # null values do exist in the dataset and df.info() indicates that column "rank"
  # is only column with them
  #============================================================================
  display("Null Values: ", df.isnull().values.any())
  '''' is any missing values in dataframe '''
  display(df.isnull())

  #============================================================================
  # lets look more closely at the rank column as we know it contains null values
  # check what unique values are stored in this column and its count
  #============================================================================
  print("\nUnique values from column rank...\n",df['rank'].unique())
  print("\nCount unique values from a column ...\n",df['rank'].nunique())

  ### get count of missing values of a particular column
 
  #display ("Count of missing values in rank col: ", df.rank.isnull().sum())
  #display (df.rank.isnull().sum())

  display("count of missing values: ", df.isnull().sum())
  display (df.isnull().sum()/len(df)*100)

  #============================================================================
  # lets look more closely at the genre column to check what values are stored 
  # in this column and its count
  #============================================================================

  print("\nUnique values from column genre...\n",df['genre'].unique())
  print("\nCount unique values from a column ...\n",df['genre'].nunique())

  #========================================================================
  # should only be 1 row per movie so check for duplicates
  #========================================================================
  duplicates = find_duplicates(df)
  #duplicates = df[df.duplicated(['name','last_name','year'])]
  #print("Duplicate Values")
  #print(duplicates)
  #print("\nCount unique values from name column ...\n",duplicates['name'].nunique())

  # check if all duplicates have missing values
  #display("count of missing values in duplicates: ", duplicates.isnull().sum())
  #display (duplicates.isnull().sum()/len(duplicates)*100)


def find_duplicates(df):
  #duplicates = df[df.duplicated(['name','last_name','year'])]
  duplicates = df[df.duplicated(['name'],keep=False)]
  display("Duplicate Values")
  display(duplicates.sort_values(by = ["name", "last_name", "year"]))
  display("\nCount unique values from name column ...\n",duplicates['name'].nunique())

  # check if all duplicates have missing values
  display("count of missing values in duplicates: ", duplicates.isnull().sum())
  display (duplicates.isnull().sum()/len(duplicates)*100)

  return duplicates 
  

# program starts here
url = "https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"

movie_df = pd.read_csv(url)

inspect_data(movie_df)

Unnamed: 0           year           rank
count  368894.000000  368894.000000  113376.000000
mean   184446.500000    1969.442222       5.995124
std    106490.669438      29.082051       1.540330
min         0.000000    1888.000000       1.000000
25%     92223.250000    1947.000000       5.100000
50%    184446.500000    1976.000000       6.200000
75%    276669.750000    1996.000000       7.100000
max    368893.000000    2007.000000       9.900000

Unnamed: 0 first_name    last_name  \
11330        11330    Mohamed         Asli   
265567      265567      Diane      Pierens   
265568      265568      Diane      Pierens   
265569      265569      Diane      Pierens   
274883      274883  Francisco  Ramalho Jr.   

                                       name  year  rank   genre  
11330    Casablanca les anges ne volent pas  2004   NaN   Drama  
265567                                Clara  1994   NaN   Drama  
265568                                Clara  1994   NaN  Family  
265569                                Clara  1994   NaN   Short  
274883                         Flor da Pele  1976   5.5   Drama

<class 'pandas.core.frame.DataFrame'>
Int64Index: 368894 entries, 11330 to 218388
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   Unnamed: 0  368894 non-null  int64  
 1   first_name  368894 non-null  object 
 2   last_name   368894 non-null  object 
 3   name        368894 non-null  object 
 4   year        368894 non-null  int64  
 5   rank        113376 non-null  float64
 6   genre       368894 non-null  object 
dtypes: float64(1), int64(2), object(4)
memory usage: 22.5+ MB


None

'Null Values: '

True

'Null Values: '

True

Unnamed: 0  first_name  last_name   name   year   rank  genre
11330        False       False      False  False  False   True  False
265567       False       False      False  False  False   True  False
265568       False       False      False  False  False   True  False
265569       False       False      False  False  False   True  False
274883       False       False      False  False  False  False  False
...            ...         ...        ...    ...    ...    ...    ...
170647       False       False      False  False  False   True  False
170648       False       False      False  False  False   True  False
218386       False       False      False  False  False   True  False
218387       False       False      False  False  False   True  False
218388       False       False      False  False  False   True  False

[368894 rows x 7 columns]


Unique values from column rank...
 [nan 5.5 8.7 5.1 5.2 7.8 6.  7.2 4.  6.8 3.1 3.9 6.2 5.8 4.3 6.4 7.3 8.1
 7.5 8.2 3.2 7.9 6.3 5.6 7.1 5.  7.  6.7 6.5 9.6 5.4 6.6 4.9 7.7 6.1 3.5
 6.9 5.7 1.6 4.6 2.  4.8 5.9 3.8 7.4 2.4 4.4 4.7 8.  3.6 5.3 2.8 4.5 2.3
 4.1 8.8 3.3 2.2 9.  7.6 9.8 8.4 9.3 4.2 9.5 8.3 1.4 3.4 2.5 3.7 1.8 9.2
 3.  8.5 2.7 2.9 9.4 1.9 8.9 8.6 9.1 9.7 2.6 1.7 1.  2.1 1.5 1.3 9.9 1.1
 1.2]

Count unique values from a column ...
 90


'count of missing values: '

Unnamed: 0         0
first_name         0
last_name          0
name               0
year               0
rank          255518
genre              0
dtype: int64

Unnamed: 0     0.000000
first_name     0.000000
last_name      0.000000
name           0.000000
year           0.000000
rank          69.265968
genre          0.000000
dtype: float64


Unique values from column genre...
 ['Drama' 'Family' 'Short' 'Comedy' 'Documentary' 'Musical' 'Romance'
 'Fantasy' 'Crime' 'Horror' 'Mystery' 'War' 'Thriller' 'Animation'
 'Western' 'Music' 'Action' 'Sci-Fi' 'Adventure' 'Film-Noir']

Count unique values from a column ...
 20


'Duplicate Values'

Unnamed: 0    first_name last_name              name  year  rank  \
265567      265567         Diane   Pierens             Clara  1994   NaN   
265568      265568         Diane   Pierens             Clara  1994   NaN   
265569      265569         Diane   Pierens             Clara  1994   NaN   
128110      128110  Pierre-Marie    Goulet             Gaule  1974   NaN   
128111      128111  Pierre-Marie    Goulet             Gaule  1974   NaN   
...            ...           ...       ...               ...   ...   ...   
170647      170647         Kemal       Kan  zmc kizin kaderi  1961   NaN   
170648      170648         Kemal       Kan  zmc kizin kaderi  1961   NaN   
218386      218386     Seiji (I)  Maruyama   zora no samurai  1976   NaN   
218387      218387     Seiji (I)  Maruyama   zora no samurai  1976   NaN   
218388      218388     Seiji (I)  Maruyama   zora no samurai  1976   NaN   

              genre  
265567        Drama  
265568       Family  
265569        Short  
128110  Documentary  
128111        Short  
...             ...  
170647        Drama  
170648      Romance  
218386       Action  
218387        Drama  
218388          War  

[270724 rows x 7 columns]

'\nCount unique values from name column ...\n'

99029

'count of missing values in duplicates: '

Unnamed: 0         0
first_name         0
last_name          0
name               0
year               0
rank          181105
genre              0
dtype: int64

Unnamed: 0     0.000000
first_name     0.000000
last_name      0.000000
name           0.000000
year           0.000000
rank          66.896544
genre          0.000000
dtype: float64



---


# **Inspecting the Data Notes**

---


1.   On initial look at the data, I can see that the dataset has **368894** rows. From  **.info()**, the only column that contains null data is the "**rank**" column. This will need to be cleaned. The first column in the dataframe is called "**Unnamed: 0**" and on initial inspection it looks like it is just a record number. Will need to look at this further.

2.   **Rank Column** - Looking more closely at the rank column which contains null values, we can see that there are **255518** records with a missing rank value and this represents nearly **70%** of the total number of recs. There are **90** unique values in this column (including nan) with a **max value of 9.9** and a **minimum of 1**. This leaves us with the question of what to do with the rank column? Options are to **a)** drop the column entirely from the dataframe **b)** somehow impute a sensible value for the empty fields **c)** remove only those records that have an empty rank value, although this would result in a dataset **reduced by nearly 70%**, a not insignificant amount.

1.  **Duplicate row check** - when checking for duplicates, no Movie/Director/Year combination should have more than 1 row. When check for rows with duplicate values, **138290** rows are returned (includes all duplicate rows) which represents **89140** rows with a unique movie name. From this initial check, it looks possible that all duplicate rows have missing values in the **rank** column. Check this out.
2.   After checking whether all duplicates have missing values in the **rank** column, I can see that this is not actually the case. Of **138290** duplicate rows returned, only **88465** of these have missing values in the rank field


2.   List item



 






### Clean the dataset 
---


In [106]:
from os import name
import pandas as pd

def clean_data(df):

  # ================================================================
  # remove all rows where column rank is null
  # ================================================================
  clean_df = df[~df["rank"].isnull()]
  #display (clean_df)

  # ================================================================
  # remove leading/trailing spaces from strings
  # ================================================================
 
  # removing whitespace from more than 1 column
  clean_df['name'].str.strip()
  clean_df['first_name'].str.strip()
  clean_df['last_name'].str.strip()

  # ================================================================
  # now that non-nulls removed, remove duplicate movie recs
  # ================================================================
  clean_duplicates = clean_df.drop_duplicates(subset = ['name'], keep = 'last')
  display ('Cleaned Duplicates: ', clean_duplicates.head(100))

  display("Sorted Duplicate Values: ")
  sort_df = clean_duplicates.sort_values(by = ["name"])
  with pd.option_context('display.max_rows', 500,):
    print("Sorted: ", sort_df)
  #display ("Sorted: ", sort_df) 
  
  display("\nCount unique values from name column ...\n",clean_duplicates['name'].nunique())
  duplicates = clean_duplicates[clean_duplicates.duplicated(['name'])]
  display ("Duplicates after cleaning: ", duplicates)


  #find_duplicates1(clean_df)
  #display (clean_df)
  
  return clean_duplicates

def find_duplicates1(df):
  #duplicates = df[df.duplicated(['name','last_name','year'])]
  duplicates = df[df.duplicated(['name'], keep = False)]

  display("Duplicate Values")
  #sort_df = duplicates.sort_values(by = ["name", "last_name"])
  #display ("Sorted: ", sort_df) 
  #display("\nCount unique values from name column ...\n",sort_df['name'].nunique())

  # check if all duplicates have missing values
  #display("count of missing values in duplicates: ", duplicates.isnull().sum())
  #display (duplicates.isnull().sum()/len(duplicates)*100)

  #groups = duplicates.groupby('name')
  #print (groups)
  #print grouped.get_group(2014)
  # group by 'X' column
  #groups = df.groupby("X")
  
  #for name, group in groups:
  #  print(name)
  #  #print(last_name)
  #  #print(first_name)
  #  print("\n")

  display (duplicates.head(50))

# program starts here
url = "https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"

movie_df = pd.read_csv(url)

clean_data(movie_df)



'Cleaned Duplicates: '

Unnamed: 0     first_name        last_name  \
77            77      Rostislav            Aalto   
78            78         Veikko         Aaltonen   
80            80         Veikko         Aaltonen   
81            81         Veikko         Aaltonen   
83            83         Veikko         Aaltonen   
85            85         Veikko         Aaltonen   
87            87         Veikko         Aaltonen   
88            88         Veikko         Aaltonen   
93            93        Douglas      Aarniokoski   
102          102           Paul            Aaron   
104          104           Paul            Aaron   
107          107           Paul            Aaron   
111          111           Paul            Aaron   
118          118           Evan         Aaronson   
127          127          Jamel         Aattache   
142          142           A.C.           Abadie   
212          212   Khwaja Ahmad            Abbas   
231          231    Abdul Malik           Abbott   
239          239    Charles (I)           Abbott   
261          261         George           Abbott   
269          269         George           Abbott   
278          278    Norman (II)           Abbott   
302          302          Daoud      Abdel Sayed   
305          305          Daoud      Abdel Sayed   
333          333          Vadim      Abdrashitov   
335          335          Vadim      Abdrashitov   
336          336          Vadim      Abdrashitov   
337          337          Vadim      Abdrashitov   
338          338          Vadim      Abdrashitov   
340          340          Vadim      Abdrashitov   
342          342          Vadim      Abdrashitov   
350          350          Aktan      Abdykalykov   
352          352          Aktan      Abdykalykov   
380          380        Tsotumo              Abe   
399          399          Ramzi             Abed   
413          413          Ramzi             Abed   
423          423  Dominique (I)             Abel   
430          430         Jeanne             Abel   
484          484            Don        Abernathy   
489          489          Lewis        Abernathy   
575          575           Brad          Ableson   
609          609           Adam          Abraham   
627          627         Derwin         Abrahams   
629          629         Derwin         Abrahams   
643          643         Derwin         Abrahams   
649          649         Derwin         Abrahams   
666          666            Jim         Abrahams   
669          669            Jim         Abrahams   
672          672            Jim         Abrahams   
676          676            Jim         Abrahams   
678          678          Lenny       Abrahamson   
684          684         A. (I)          Abramov   
736          736           Neil         Abramson   
739          739        Aluisio        Abranches   
740          740        Aluisio        Abranches   
747          747       Gilda de            Abreu   
755          755           Levi           Abrino   
777          777           Hany        Abu-Assad   
779          779           Hany        Abu-Assad   
785          785           Jeff           Abugov   
792          792         Tengiz         Abuladze   
793          793         Tengiz         Abuladze   
794          794         Tengiz         Abuladze   
854          854            Jon          Acevski   
865          865           Mark           Achbar   
873          873         Stuart            Acher   
879          879    Nan Triveni           Achnas   
887          887    Nan Triveni           Achnas   
891          891           Irma           Achten   
895          895        Herbert     Achternbusch   
903          903          Dejan         Acimovic   
913          913          Karpo  Acimovic-Godina   
915          915          Karpo  Acimovic-Godina   
923          923          Jovan             Acin   
938          938          Shane            Acker   
950          950      David (I)         Ackerman   
96

'Sorted Duplicate Values: '

Sorted:          Unnamed: 0 first_name     last_name  \
274883      274883  Francisco   Ramalho Jr.   
233647      233647     Evaldo      Mocarzel   
233670      233670   Giuseppe        Moccia   
118878      118878     Pierre  Gaspard-Huit   
216153      216153      Mario        Marret   
...            ...        ...           ...   
81940        81940    Jacques          Demy   
1900          1900     Amable       Aguiluz   
20127        20127    Mariano       Barroso   
198580      198580     Gianni         Lepre   
151291      151291     Marius         Holst   

                                                     name  year  rank  \
274883                                       Flor da Pele  1976   5.5   
233647                                   Margem da Imagem  2003   8.7   
233670                              arrivato mio fratello  1985   5.1   
118878                                       belles dents  1966   5.5   
216153                                    bientt, j'espre  19

'\nCount unique values from name column ...\n'

55954

'Duplicates after cleaning: '

Empty DataFrame
Columns: [Unnamed: 0, first_name, last_name, name, year, rank, genre]
Index: []

Unnamed: 0 first_name last_name                                name  \
77              77  Rostislav     Aalto                        Cleaning Up!   
78              78     Veikko  Aaltonen                            Is meidn   
80              80     Veikko  Aaltonen                     Juoksuhaudantie   
81              81     Veikko  Aaltonen                                 Maa   
83              83     Veikko  Aaltonen                          Merisairas   
...            ...        ...       ...                                 ...   
368829      368829     Memduh        Ün                            Cevriyem   
368833      368833     Memduh        Ün                         Devlet Kusu   
368838      368838     Memduh        Ün                               Garip   
368862      368862     Memduh        Ün              Sezercik aslan parcasi   
368891      368891       Idil      Üner  Liebenden vom Hotel von Osman, Die   

        year  rank        genre  
77      2002   8.3  Documentary  
78      1993   6.8        Drama  
80      2004   6.3        Drama  
81      2001   6.0  Documentary  
83      1996   4.3     Thriller  
...      ...   ...          ...  
368829  1978   7.8      Romance  
368833  1980   7.2        Drama  
368838  1986   7.9        Drama  
368862  1972   6.0      Romance  
368891  2001   8.5        Short  

[55954 rows x 7 columns]

In [95]:
from os import name
import pandas as pd

def misc_data(df):
  return


# program starts here
url = "https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"

movie_df = pd.read_csv(url)
#df3 = movie_df[(movie_df['last_name'] == 'Aaltonen') & (movie_df['first_name'] == 'Veikko')]
df3 = movie_df
#display (df3)

# ================================================================
# remove all rows where column rank is null
# ================================================================
clean_df = df3[~df3["rank"].isnull()]
#display (clean_df)

# ================================================================
# check for duplicate movies
# ================================================================
clean_df = df3[~df3["rank"].isnull()]
duplicates = clean_df[clean_df.duplicated(['name'],keep=False)]
display ("Duplicates: ", duplicates)
#print(duplicates.to_markdown())
# The scope of these changes
# are local with systems to with statement.
with pd.option_context('display.max_rows', 500,):
    print(duplicates)

display ("Aaltonen Duplicates: ", duplicates[(duplicates['last_name'] == 'Aaltonen')])
display ("Ün Duplicates: ", duplicates[(duplicates['last_name'] == 'Ün')])


misc_data(movie_df)

'Duplicates: '

Unnamed: 0 first_name last_name                    name  year  rank  \
79              79     Veikko  Aaltonen         Juoksuhaudantie  2004   6.3   
80              80     Veikko  Aaltonen         Juoksuhaudantie  2004   6.3   
82              82     Veikko  Aaltonen              Merisairas  1996   4.3   
83              83     Veikko  Aaltonen              Merisairas  1996   4.3   
84              84     Veikko  Aaltonen      Rakkaudella, Maire  1999   6.9   
...            ...        ...       ...                     ...   ...   ...   
368833      368833     Memduh        Ün             Devlet Kusu  1980   7.2   
368837      368837     Memduh        Ün                   Garip  1986   7.9   
368838      368838     Memduh        Ün                   Garip  1986   7.9   
368861      368861     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   
368862      368862     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   

           genre  
79        Comedy  
80         Drama  
82         Drama  
83      Thriller  
84         Drama  
...          ...  
368833     Drama  
368837    Comedy  
368838     Drama  
368861     Drama  
368862   Romance  

[87669 rows x 7 columns]

        Unnamed: 0 first_name last_name                    name  year  rank  \
79              79     Veikko  Aaltonen         Juoksuhaudantie  2004   6.3   
80              80     Veikko  Aaltonen         Juoksuhaudantie  2004   6.3   
82              82     Veikko  Aaltonen              Merisairas  1996   4.3   
83              83     Veikko  Aaltonen              Merisairas  1996   4.3   
84              84     Veikko  Aaltonen      Rakkaudella, Maire  1999   6.9   
...            ...        ...       ...                     ...   ...   ...   
368833      368833     Memduh        Ün             Devlet Kusu  1980   7.2   
368837      368837     Memduh        Ün                   Garip  1986   7.9   
368838      368838     Memduh        Ün                   Garip  1986   7.9   
368861      368861     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   
368862      368862     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   

           genre  
79        Comedy  
80         Dr

'Aaltonen Duplicates: '

Unnamed: 0 first_name last_name                name  year  rank     genre
79          79     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3    Comedy
80          80     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3     Drama
82          82     Veikko  Aaltonen          Merisairas  1996   4.3     Drama
83          83     Veikko  Aaltonen          Merisairas  1996   4.3  Thriller
84          84     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9     Drama
85          85     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9  Thriller
86          86     Veikko  Aaltonen           Tilinteko  1987   6.0     Crime
87          87     Veikko  Aaltonen           Tilinteko  1987   6.0     Drama

'Ün Duplicates: '

Unnamed: 0 first_name last_name                    name  year  rank  \
368828      368828     Memduh        Ün                Cevriyem  1978   7.8   
368829      368829     Memduh        Ün                Cevriyem  1978   7.8   
368832      368832     Memduh        Ün             Devlet Kusu  1980   7.2   
368833      368833     Memduh        Ün             Devlet Kusu  1980   7.2   
368837      368837     Memduh        Ün                   Garip  1986   7.9   
368838      368838     Memduh        Ün                   Garip  1986   7.9   
368861      368861     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   
368862      368862     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   

          genre  
368828    Drama  
368829  Romance  
368832   Comedy  
368833    Drama  
368837   Comedy  
368838    Drama  
368861    Drama  
368862  Romance



---


# **Cleaning the Dataset Notes**


---



1.  **Missing Values** - Column **rank** is the only col with missing values. It is a vital piece of info for our data analysis so I'm left with the choice somehow imputing a sensible value into this col or deleting all rows where it has null values or it is non-numeric. As there is no means to impute a sensible value, I have decided that I will need to disregard all rows where rank is not present in all further analysis.

2.   **Cleaning duplicates** - logically there should be 1 record per movie in the dataset, however looking at the data I can see that there are duplicates



### How many movies of each genre are there?
---

In [ ]:
from os import name
import pandas as pd

def movies_by_genre():
  
  return

# program starts here
movies_by_genre(movie_df)

  

### Which director has the highest ranked movies?
---

### How many movies have ranks of over 9?
---


### Plot a bar chart of mean rank and genre
---

### Plot a pie chart of how many movies of each genre there are 
---


### Plot a graph showing the mean Rank for each year 

### What else can you find out from this dataset?
---
Make a plan of 3 further things you can do to interrogate and analyse this dataset

Type your answer here  

1.  
2.  
3.   

### Complete the tasks you have set out in the exercise above. 
---


# Reflection
----

## What skills have you demonstrated in completing this notebook?

Your answer:

## What caused you the most difficulty?

Your answer: